Rough draft of Section 3.5 conditionals: #if, #elif, etc. (#61)
j3-fortran/fpp-202y: drafts/25-xxx-specifications.txt

di04. The '#include' directive

di05. The '#if', '#ifdef', '#ifndef', '#elif', '#elifdef',
      '#elifndef', '#else', '#endif' conditional directives

di06. The '#error' and '#warning' directives

di07. The '#line' directive

di08. The '#pragma' directive

> 

Reviewer: Stray blank lines?

Author: Fixed in new push.

      the #include directive is not prohibited from including the
      same source file at a deeper level of nesting.

3.5 The '#if', '#ifdef', '#ifndef', '#elif', '#elifdef',
      '#elifndef', '#else', '#endif' conditional directives
---------------------------------------------------------

Example syntax (extra spacing for illustration purposes only):
    General form:

>         <if-head>  <lines>                        #endif EOL
>         <if-head>  <lines>  <if-elif>             #endif EOL
>         <if-head>  <lines>  <if-elif>  <if-else>  #endif EOL

Reviewer 2: Duplicate `<lines>` in `<if-head>`
```suggestion
        <if-head>                        #endif EOL
        <if-head>  <if-elif>             #endif EOL
        <if-head>  <if-elif>  <if-else>  #endif EOL
```

Author: Thanks. I missed this when refactoring the grammar on the fly.

    Where:
        <if-head> is one of:
            #if token-list EOL  <lines>
            #ifdef ID EOL       <lines>
            #ifndef ID EOL      <lines>

        <if-elif> is zero or more of:
            #elif     token-list EOL  <lines>
            #elifdef  ID EOL          <lines>
            #elifndef ID EOL          <lines>

        <if-else> is zero or one of:
            #else EOL  <lines>

        <lines> are zero or more of:
            directive lines
            Fortran comment lines
            Fortran source lines

3.5.1 Static semantics specifications
-------------------------------------

> if05. A #if, ##ifdef, or #ifndef directive signals the start of a list

Reviewer 2: ```suggestion
if05. A #if, #ifdef, or #ifndef directive signals the start of a list
```

Reviewer 3: Ha, I saw that one too, and was afraid it was something from a meeting I missed.

      of conditional directives.

if10: Whitespace must immediately follow the directives tokens
      'ifdef', 'ifndef', 'elifdef', and 'elifndef' to separate them

>       from the ID tokens.

Reviewer 2: This property is roughly true for (nearly?) every preprocessing directive, not even just those in this section. Stating it here for only three makes them seem like an exception. (I say roughly because what's actually required is a token boundary, which might or might not include whitespace)

I think what we actually _mean_ here is that standard directives are only recognized as such if the token immediately following the `#` introducing a directive line exactly matches one of the standard directive names. If it does not match for any reason (e.g. due to lack of whitespace) then the directive does not conform to one of the standard directives and is instead parsed as a processor-dependent directive.

We should probably directly state something to this effect in section 3.0 and then delete this rule.

Reviewer 3: Section 2.3 is titled "Significance of whitespace" - would this fit there?

Author: I thought I replied that Section 2.3 seems like an excellent place for it. I will generate a pull request for section 2 for this.

Author: Added text in today's update. Will request further reviews.

if15. The list of #elif, #elifdef, and #elifndef directives, and the
      #else directive (if present) are part of the same list of
      directives introduced by the #if directive.

if20. The chain of conditional directives ends with the #endif
      directive.

if25. Within a conditional directive list, a #if, #ifdef, or #endif
      directive introduces a new list of conditional directives, at a new
      nesting level, within the enclosing conditional directive list.

if30. The conditional directive lists must properly nest.

if35. As shown in the illustrative syntax, there should be no

>       unmatched #endif directives.

Reviewer 2: This property can be directly derived from the pseudo-grammar above, so repeating it here is redundant. Worse, we only state one such property but omit others, for example: 

1. there should be no `<if-head>` without a matching `#endif`, 
2. there should be no `<if-elif>` without a matching `<if-head>` and `#endif`, 
3. there should be no `#else` without a matching `<if-head>` and `#endif`, 

which makes this one seem like an outlier. We should either list all such analogous grammatical properties or none.

Author: I'm OK with letting the grammar speak for itself here, as long as we believe J3 as a whole won't feel like it's under-specified.

Author: Remove if35.

Author: Made changes in today's update. Will request further reviews.

if40. As shown in the illustrative syntax, there should be no more
      than one #else directive in a list of conditional directives at
      any nesting level.

3.5.2 Evaluation semantics specifications
-----------------------------------------

> ix05. The #ifdef and #ifndef macros are evaluated as if they had been

Reviewer 2: ```suggestion
ix05. The #ifdef and #ifndef directives are evaluated as if they had been
```

      written "#if defined(ID)" and "#if !defined(ID)" respectively.
      (For brevity, the descriptions of #ifdef and #ifndef directives
      below are omitted, and assume this transformation.)

ix10. The #elifdef and #elifndef macros are evaluated as if they had
      been written "#elif defined(ID)" and "#elif !defined(ID)"
      respectively.

ix15. The token-lists in #if and #elif are processed as described in

>       the section "Macro recognition and expansion".

Reviewer 2: As noted in the meeting, we also need to clarify whether macro expansion occurs within an `#elif` directive line if we already know that "group" is being skipped

Author: Added text in today's update. Will request further reviews.

ix20. After expansion, the resulting token list in #if and #elif
      directives must represent a valid integer expression as
      described in the section "Expressions allowed in #if and #elif
      directives". This expression is called the "controlling
      expression" for the directive.

ix25. Controlling expressions in #if and #elif directives are parsed
      and evaluated according to the semantic rules described in
      section "Expressions allowed in #if and #elif directives".

> ix30. If the controlling expression in a #if evaluates to a nonzero
>       value, the <lines> immediately following the #if participate in
>       preprocessing. They are subject to preprocessing expansion,
>       substitution, and evaluation of processing directives in the
>       <lines>.
> 
> ix40. Further preprocessing when #if controlling expression evaluates
>       to a nonzero value proceeds. Subsequent #elif and #else
>       directives at the same level of nesting are skipped, and do not
>       participate in preprocessing except to be examined for proper
>       nesting of conditional directives.

Reviewer 2: I'm unhappy with the structure of how we describe the controlled `<lines>` here and in the subsequent rules. 

In particular, the definition of phrases like "<lines> immediately following the #if" and "<lines> immediately following the #elif" is too ambiguous, especially if those `<lines>` themselves contain directives (and possibly nested conditionals).

I think we should instead use the grammar non-terminals to help make this more precise. For example:

```suggestion
ix30. If the controlling expression in an <if-head> construct evaluates to a nonzero
      value, then the <lines> contained within that <if-head> construct participate in
      preprocessing. They are subject to preprocessing expansion,
      substitution, and evaluation of processing directives in those
      <lines>.

ix40. Further preprocessing when an <if-head> controlling expression evaluates
      to a nonzero value proceeds. Subsequent <if-elif> and <if-else>
      constructs at the same level of nesting are skipped, and the <lines>
      contained within those constructs do not participate in preprocessing 
      except to be examined for proper nesting of conditional directives.
```
and similarly for the next four rules.

Thoughts?

Author: I like these updates to the phrasing. I am OK committing the phrasing for this one as suggested, and then updating the PR with new commits to add changes for the remaining rules.

Author: We will make these changes. And others, regarding use of `[ ]` for optional things, and _-list_ for lists.

Author: Changes have been made for this in today's update. Will request further reviews.

ix45. If the controlling expression in a #if evaluates to a zero
      value, the lines immediately following the #if directive are
      skipped, and examined only to determine nesting of conditionals.
      Within these lines, no token expansion occurs, and no Fortran
      comment lines nor source lines are made available for subsequent

>       processing by the processor. Preprocessing continues with any

Reviewer: Do we want to specifically call out that macros in the body are not evaluated in any way, in addition to the lack of expansion and source lines? I'm thinking of things like `#error`, `#warning`, and `#pragma` which aren't really about token expansion or source lines.

Author: You're right. We should be clear that in the skipped lines,

- Nested conditional directives are recognized but no evaluated
- No other directives are evaluated
- No macro expansion occurs
- No directive lines, Fortran source lines, or Fortran comment lines are made available for subsequent processing

Author: Make sure that skipped parts act as if they aren't there, and have no effect on the preprocessing.

Author: Added text in today's update. This caused all the subsequent specifications on "skipping" and "participating in further preprocessing" to change. Will request further reviews.

        <if-head>                             #endif EOL
        <if-head>  <if-elif-list>             #endif EOL
        <if-head>  <if-elif-list>  <if-else>  #endif EOL

    Where:
        <if-head> is one of:
            #if token-list EOL  <group>
            #ifdef ID EOL       <group>
            #ifndef ID EOL      <group>

        <if-elif-list> is zero or more of:
            <if-elif> <group>

        <if-elif> is one of:
            #elif     token-list EOL  <group>
            #elifdef  ID EOL          <group>
            #elifndef ID EOL          <group>

        <if-else> is one of:
            #else EOL  <group>

        <group> is zero or more of:
            directive lines
            Fortran comment lines
            Fortran source lines

3.5.1 Static semantics specifications
-------------------------------------

if05. A #if, #ifdef, or #ifndef directive signals the start of a list
      of conditional directives.

if15. The list of #elif, #elifdef, and #elifndef directives, and the
      #else directive (if present) are part of the same list of
      directives introduced by the #if directive.

if20. The chain of conditional directives ends with the #endif

>       directive.

Reviewer 2: RESOLVED in 05-28 meeting: rename all instances to chain
